### PR TITLE
Fix skip last calculation at increased playback speeds

### PR DESCRIPTION
**Pull request by sztomek**
## Description
Skip Last was not accounting for playback speed: the trigger compared media-time remaining against the configured `skipLastSecs`, so a user playing at 2x with Skip Last set to 2:00 only saved ~1:00 of listening time before the episode auto-skipped. This was reported via Zendesk #11111728 and confirmed by the reporter's own investigation at 2x speed.                                                                                                                                                                                                                                                                               
This branch extracts two pure helpers (`effectiveSkipLastMs` and `shouldSkipLast`) on `PlaybackManager`'s companion object and multiplies the threshold by the current playbackSpeed, so the setting now represents listening time regardless of speed. Sub-1x speeds are clamped to 1x so slow-playback users never get a shorter skip window than they configured, and the log line now includes playbackSpeed to make future support tickets easier to diagnose.                                                                                                                                                                                       
At 1.0x the behaviour is identical to before, so users who don't change speed see no change. New unit tests lock in the 1x baseline, verify the 2x trigger at the correct listening-time threshold, cover the sub-1x guard, and guard the null/zero edge cases.       

Fixes PCDROID-542 https://linear.app/a8c/issue/PCDROID-542/skip-last-fires-too-early-at-higher-playback-speeds-skipping-less

## Testing Instructions
1. Install the debug build on a device                                                                                                                                                                                                                         2. Subscribe to any podcast with episodes longer than ~10 minutes                                                                                                     
3. Open the podcast settings and set Skip Last to 2:00.                                                                                                                                                                                                                                                                 
4. Open the global playback effects and set Playback speed to 2.0x.                                                                                                                                                                                                                                                      5. Play an episode and seek to near the end; confirm the auto-skip fires when the player's displayed remaining time reads ~2:00 (previously it fired at ~1:00). A "Skipped last 2 minutes" toast should appear.                                                                                                
6. Change the playback speed to 1.0x and repeat — Skip Last should fire at ~2:00 remaining (unchanged behaviour).                                                                                                                                                                                                       
7. Change the playback speed to 0.8x and repeat — Skip Last should still fire at ~2:00 remaining, never earlier (sub-1x guard).                                                                                                                                                                                       
8. Set Skip Last to 0 (off) and confirm the episode plays to its natural end with no auto-skip, at any playback speed.       


## Checklist
- [x] If this is a user-facing change, I have added an entry in CHANGELOG.md
- [x] Ensure the linter passes (`./gradlew spotlessApply` to automatically apply formatting/linting)
- [x] I have considered whether it makes sense to add tests for my changes
- [x] All strings that need to be localized are in `modules/services/localization/src/main/res/values/strings.xml`
- [x] Any jetpack compose components I added or changed are covered by compose previews
- [x] I have updated (or requested that someone edit) [the spreadsheet](https://docs.google.com/spreadsheets/d/107jqrutZhU0fVZJ19SBqxxVKbV2NWSdQC9MFYdLiAxc/edit?usp=sharing) to reflect any new or changed analytics.

### CODE REVIEW

File: modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/PlaybackManager.kt [hunk 1]
```diff
@@ -169,6 +169,25 @@ open class PlaybackManager @Inject constructor(
...
+            playbackSpeed: Double,
+            positionMs: Int,
+            durationMs: Int?,
+        ): Boolean {
+            if (skipLastSecs == null || skipLastSecs <= 0) return false
+            if (durationMs == null || durationMs <= 0) return false
+            if (positionMs < 0) return false
+            val thresholdMs = effectiveSkipLastMs(skipLastSecs, playbackSpeed)
+            if (durationMs <= thresholdMs) return false
+            return durationMs - positionMs < thresholdMs
```
[the remark this answers is not in the document]
Comment (sztomek) [reply modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/PlaybackManager.kt:191]: Kotlin's `Int` has built-in `compareTo(Long)` overloads, so `durationMs <= thresholdMs` and `durationMs - positionMs < thresholdMs` compile and work correctly — no explicit `.toLong()` conversion needed. The tests confirm this compiles and passes.

**Comment by geekygecko** [comment]
@sztomek would you mind looking at the comments above?

File: modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/PlaybackManager.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (sztomek) [reply modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/PlaybackManager.kt:191]: Already addressed in a previous commit — Kotlin's `Int` has built-in `compareTo(Long)` overloads, so no explicit conversion is needed.

File: modules/services/repositories/src/test/java/au/com/shiftyjelly/pocketcasts/repositories/playback/PlaybackManagerSkipLastTest.kt [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    @Test
+    fun `shouldSkipLast at 0 point 5x keeps the 1x window (never shrinks)`() {
+        // The user set Skip Last=120s. On 0.5x we must not reduce the media window below 120_000ms.
+        // At 60_001ms remaining (position 1_739_999), 1x would NOT fire, so neither should 0.5x.
+        assertFalse(
+            shouldSkipLast(
+                skipLastSecs = 120,
+                playbackSpeed = 0.5,
+                positionMs = 1_679_999,
+                durationMs = 1_800_000,
```
[the remark this answers is not in the document]
Comment (sztomek) [reply modules/services/repositories/src/test/java/au/com/shiftyjelly/pocketcasts/repositories/playback/PlaybackManagerSkipLastTest.kt:108]: This was fixed in a previous commit — the comment now correctly reads "120_001ms remaining (position 1_679_999)" which matches the test values.

File: modules/services/repositories/src/test/java/au/com/shiftyjelly/pocketcasts/repositories/playback/PlaybackManagerSkipLastTest.kt [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        assertFalse(
+            shouldSkipLast(skipLastSecs = 120, playbackSpeed = 1.0, positionMs = -1, durationMs = 1_800_000),
+        )
+    }
+
+    @Test
+    fun `shouldSkipLast returns false when episode is shorter than the skip-last window`() {
+        // A 60s episode with skipLast=120s at 1x: the whole episode is inside the window, but we
+        // intentionally do NOT fire because the guard `durationMs > thresholdMs` filters it out.
+        // (Preserves prior behaviour for very short episodes.)
```
[the remark this answers is not in the document]
Comment (sztomek) [reply modules/services/repositories/src/test/java/au/com/shiftyjelly/pocketcasts/repositories/playback/PlaybackManagerSkipLastTest.kt]: Good catch — the old code compared `durationMs` (ms) against `skipLast` (seconds), so it never actually filtered short episodes correctly. Updated the comment to describe the intended behaviour without claiming it preserves prior behaviour.

File: modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/PlaybackManager.kt [hunk 4]
```diff
@@ -2350,35 +2369,35 @@ open class PlaybackManager @Inject constructor(
             return
         }
 
-        val skipLast = playbackStateRelay.blockingFirst().podcast?.skipLastSecs
+        val currentPlaybackState = playbackStateRelay.blockingFirst()
+        val skipLast = currentPlaybackState.podcast?.skipLastSecs
+        val playbackSpeed = currentPlaybackState.playbackSpeed
```
[the remark this answers is not in the document]
Comment (sztomek) [reply modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/PlaybackManager.kt:2383]: Good point — consolidated to a single `blockingFirst()` call and reused for both the episode UUID check and reading `skipLast`/`playbackSpeed`.

File: modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/PlaybackManager.kt [hunk 5]
```diff
@@ -170,6 +170,25 @@ open class PlaybackManager @Inject constructor(
         private const val MAX_TIME_WITHOUT_FOCUS_FOR_RESUME_MINUTES = 30
         private const val MAX_TIME_WITHOUT_FOCUS_FOR_RESUME = (MAX_TIME_WITHOUT_FOCUS_FOR_RESUME_MINUTES * 60 * 1000).toLong()
         private const val PAUSE_TIMER_DELAY = ((MAX_TIME_WITHOUT_FOCUS_FOR_RESUME_MINUTES + 1) * 60 * 1000).toLong()
+
+        internal fun effectiveSkipLastMs(skipLastSecs: Int?, playbackSpeed: Double): Long {
+            val secs = skipLastSecs ?: 0
+            return (secs * 1000L * playbackSpeed.coerceAtLeast(1.0)).toLong()
```
[the remark this answers is not in the document]
Comment (sztomek) [reply modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/PlaybackManager.kt]: Valid concern — `NaN.coerceAtLeast(1.0)` stays `NaN` since NaN comparisons always return false. Added `takeIf { it.isFinite() } ?: 1.0` before the clamp, plus a test for the NaN case.

File: modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/PlaybackManager.kt [hunk 6]
```diff
@@ -2350,35 +2369,35 @@ open class PlaybackManager @Inject constructor(
...
+                episodeManager.markAsPlayedBlocking(episode, this, podcastManager)
+            } else {
+                eventHorizon.track(
+                    PlayerEpisodeCompletedEvent(
+                        podcastUuid = episode.podcastOrSubstituteUuid,
+                        episodeUuid = episode.uuid,
+                    ),
+                )
+                statsManager.addTimeSavedAutoSkipping(remainingMs.toLong())
+                episodeManager.markAsPlayedBlocking(episode, this, podcastManager)
```
[the remark this answers is not in the document]
Comment (sztomek) [reply modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/PlaybackManager.kt:2405]: Good catch — added `.coerceAtLeast(0)` on the `remainingMs` calculation to prevent subtracting from the saved-time stat if the player position overshoots the duration.

**Comment by sztomek** [comment]
@geekygecko addressed the comments

**Comment by sztomek** [comment]
@geekygecko no pressure, but did you have a chance to review the code?

**Comment by geekygecko** [comment]
Thanks for addressing the comments, I will take a look today.

**Review by geekygecko — COMMENTED** [review]
I could be misunderstanding, but I’m not sure this change makes sense to me.

I thought the idea of ‘Skip Last’ was that a podcast regularly has a segment at the end of each episode that is always the same length. The user’s playback speed shouldn’t change that duration.

I tested this with ‘Skip First’ set to 60 seconds and playback speed set to 2x, and it still only skipped the first 60 seconds for me.
